### PR TITLE
perf(ir): speed up expr and op equality comparisons by caching

### PR DESCRIPTION
**Pull request by cpcloud**
This PR refactors the caching mechanism for expr and op equality comparison. Previously we were caching in a very ad hoc manner. After this PR, subclasses of util.CachedEq need only implement the `_raw_equals` interface to gain the benefits of caching. Additionally, some comparisons have been specifically implemented to hit the short circuit path as soon as possible.

### CODE REVIEW

File: ibis/expr/datatypes/__init__.py [hunk 1]
```diff
@@ -82,11 +84,17 @@ def _factory(self, nullable: bool = True) -> DataType:
...
-        return self.equals(other)
-
-    def __ne__(self, other) -> bool:
-        return not (self == other)
+    def equals(
+        self,
+        other: typing.Any,
+        cache: MutableMapping[Hashable, bool] | None = None,
+    ) -> bool:
+        if not isinstance(other, DataType):
```
Comment (kszucs) [inline ibis/expr/datatypes/__init__.py:92]: Can we move this check to `CachedEq`?

File: ibis/util.py [hunk 2]
```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
...
     return decorator
+
+
+class Eq:
+    """A mixin for abstracting equality checks."""
+
+    __slots__ = ()
+
+    def __eq__(self, other: Any) -> bool:
+        return self.equals(other, cache={})
```
Comment (kszucs) [inline ibis/util.py:418]: How about 

```py
try:
    return self.equals(other, cache={})
except:
    raise NotImplemented
```

?

File: ibis/util.py [hunk 3]
```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
...
+
+    def __ne__(self, other: Any) -> bool:
+        return not (self == other)
+
+    def equals(
+        self,
+        other: Any,
+        cache: MutableMapping[Hashable, bool] | None = None,
+    ) -> bool:
+        return self is other or (
```
Comment (kszucs) [inline ibis/util.py:428]: Along with my previous suggestion:

```python
if self is other:
    return True
if type(self) != type(other):
    raise TypeError(...)
return self._raw_equals(other, cache=cache)
```

File: ibis/util.py [hunk 4]
```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     return decorator
+
+
+class Eq:
```
Comment (kszucs) [inline ibis/util.py:412]: Can we call it either `EqMixin` or some `*able`, like `Comparable`?

File: ibis/util.py [hunk 5]
```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
...
+        cache: MutableMapping[Hashable, bool] | None = None,
+    ) -> bool:
+        return self is other or (
+            type(self) == type(other) and self._raw_equals(other, cache=cache)
+        )
+
+    def _raw_equals(
+        self,
+        other: Any,
+        cache: MutableMapping[Hashable, bool],
```
Comment (kszucs) [inline ibis/util.py:435]: Turn it into an `@abstractmethod`?

File: ibis/util.py [hunk 6]
```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
...
+    ) -> bool:
+        """Compare the individual fields of a subclass."""
+        return True
+
+
+class CachedEq(Eq):
```
Comment (kszucs) [inline ibis/util.py:441]: How about merging these classes? `cache` arg in `Eq._raw_equals`'s signature seems odd.

File: ibis/util.py [hunk 7]
```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
...
+    def equals(
+        self,
+        other: Any,
+        cache: MutableMapping[Hashable, bool] | None = None,
+    ) -> bool:
+        return self is other or (
+            type(self) == type(other) and self._raw_equals(other, cache=cache)
+        )
+
+    def _raw_equals(
```
Comment (kszucs) [inline ibis/util.py:432]: Perhaps `__equals__` or that would be too confusing?

Comment (cpcloud) [reply ibis/util.py:432]: How about `__component_eq__`?

File: ibis/util.py (hunk 6 above)
Comment (cpcloud) [reply ibis/util.py:441]: They are separate to allow `Annotable` subclasses that aren't hashable. I'm not sure whether this occurs in practice but there are a few tests that use them.

Comment (kszucs) [reply ibis/util.py]: That shouldn't be possible asof the immutable enforcement PRs land.

Comment (cpcloud) [reply ibis/util.py]: Makes sense, let's land that first then.

**Comment by cpcloud** [comment]
@kszucs Thoughts on this one?

File: ibis/util.py (hunk 7 above)
Comment (kszucs) [reply ibis/util.py]: Actually __field_eq__ used at the datatypes looks like a better name. It should make the port of datatypes to be annotable easier too.

File: ibis/util.py (hunk 6 above)
Comment (kszucs) [reply ibis/util.py]: Possibly requires a rebase but on top of that PR

**Comment by cpcloud** [comment]
@jreback Actually if you could review this one, we should land it before #3613.

File: ibis/expr/datatypes/__init__.py [hunk 8]
```diff
@@ -82,11 +85,17 @@ def _factory(self, nullable: bool = True) -> DataType:
...
+            raise TypeError(
+                'Comparing datatypes to other types is not allowed. Convert '
```
Comment (jreback) [inline ibis/expr/datatypes/__init__.py:95]: shouldn't this return False? and not raise

File: ibis/expr/datatypes/__init__.py [hunk 9]
```diff
@@ -702,6 +790,16 @@ def __init__(
         self.value_type = dtype(value_type)
         self._pretty += f"<{self.value_type}>"
 
+    def _literal_value_hash_key(self, value):
+        return self, _tuplize(value)
+
+    def __fields_eq__(
+        self,
+        other: Set,
```
Comment (jreback) [inline ibis/expr/datatypes/__init__.py:798]: can u put some of these in a super class?

File: ibis/expr/operations/relations.py [hunk 10]
```diff
@@ -743,6 +935,19 @@ class ExistsSubquery(Node):
...
+    def __component_eq__(
+        self,
+        other: ExistsSubquery,
+        cache: MutableMapping[Hashable, bool],
+    ) -> bool:
+        return self.foreign_table.equals(
+            other.foreign_table,
+            cache=cache,
+        ) and all(
+            left_pred.equals(right_pred, cache=cache)
```
Comment (jreback) [inline ibis/expr/operations/relations.py:947]: maybe with a helper function to avoid writing all of the left right equals repeatedly

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/expr/datatypes/__init__.py (hunk 9 above)
Comment (cpcloud) [reply ibis/expr/datatypes/__init__.py:798]: No, the point is to ensure that any subclass of `DataType` defines it and avoid any implicit behavior.

File: ibis/expr/datatypes/__init__.py (hunk 8 above)
Comment (cpcloud) [reply ibis/expr/datatypes/__init__.py:95]: No, comparing a datatype with something other than a datatype is an error.

File: ibis/expr/datatypes/__init__.py (hunk 9 above)
Comment (jreback) [reply ibis/expr/datatypes/__init__.py:798]: got it makes sense

File: ibis/expr/operations/relations.py (hunk 10 above)
Comment (cpcloud) [reply ibis/expr/operations/relations.py]: Done. Added `util.seq_eq` for comparing sequences of `CachedEqMixin` subclasses.

**Comment by cpcloud** [comment]
@jreback This should be good to go on green.

**Review by jreback — APPROVED** [review]

**Pull request merged**